### PR TITLE
Abstract flexbox over LayoutTree (rebased on top of latest main)

### PR DESCRIPTION
**Pull request by nicoburns**
This PR is #202 rebased onto the history rewrite, and with the latest main merged in to resolve conflicts.

Files other than flexbox.rs were merged automatically by git.
Flexbox.rs was taken from this branch, and the following commits were manually ported in:

- Make library more usable in a `const` context (#203)
- Move `hidden_layout()` to parent scope (#208)
- Move flex-item and flex-line layout passes to methods instead of closures (#209)
- Add support for align-items: space-evenly

Original PR description below:

---

This PR makes the flexbox algorithm generic over a LayoutTree trait.

At the time of filing this PR, the LayoutTree trait mostly abstracts over the broadest set of functions that the Taffy struct satisfies. I don't know or think this is the best and most general solution to flexboxing: Taffy internally does caching of calculations that a user's LayoutTree shouldn't really be concerned about. Perhaps some of the state - like caching - could remain in a Taffy struct, but this level of specificity intrinsically ties a user's UI system to Taffy due to dependent data structures.

Generally, since Taffy uses its own types, most implementations will be tied to Taffy anyways, so perhaps this linking of user code to Taffy is not a big deal.

Also - this trait supplies a method that returns a FlexboxLayout (not the final layout, despite the confusing name) which might not be exactly a trait implementation would return the "current state" of the DOM Node. For instance, in Dioxus, we are using the CSS Cascade to supply the "current state" of the DOM node's style. I think before publishing out these changes, it might be worth further abstracting the trait to supply single-level Taffy attributes. Perhaps a "TaffyAttribute" enum could be supplied so use cases like Dioxus' cascading CSS could be implemented efficiently.

Closes https://github.com/DioxusLabs/taffy/issues/182

### CODE REVIEW

**Comment by nicoburns** [comment]
@alice-i-cecile What would it take to get this PR merged?
(note: this PR is not my work. I just cleaned up the git conflicts because I would like to build on top of it)

While I don't think the abstraction has quite been perfected yet, I'd rather do further work on top of this (and make further improvements in follow up PRs), because I think it's much cleaner (key improvement for me is passing the storage in an explicit parameter hidden behind a trait, rather than using self). However, I'm also reluctant to build too much on top of this as:

- They are then also blocked on this being merged
- It's quite a big refactor and it seems to generate not-particularly-easy-to-resolve conflicts every time other PRs are merged.

**Comment by alice-i-cecile** [comment]
I need to dedicate an hour or two to review this, but I similarly feel that this should be merged pretty aggressively.

I'll try to get that done for you by end of day next Monday (my 20% time at work). May happen before that, but you can hold me to that deadline.

File: README.md [hunk 1]
```diff
@@ -13,6 +13,44 @@ Right now, it powers:
 - [Dioxus](https://dioxuslabs.com/): a React-like library for building fast, portable, and beautiful user interfaces with Rust
 - [Bevy](https://bevyengine.org/): an ergonomic, ECS-first Rust game engine
 
+## Usage
+
+### Through the LayoutTree trait
+
+In order to be more flexible over the underlying storage, taffy provides the `LayoutTree` trait. This trait assumes that your implementation will be storing Taffy-specific data, like style and layout information. For the `Flexbox` algorithm, Taffy expects that your implementation of this trait will provide Flexbox-specific data, like its flex rules and specified dimensions.
```
Comment (alice-i-cecile) [inline README.md:20]:
```suggestion
In order to be more flexible over the underlying storage, Taffy provides the `LayoutTree` trait. This trait assumes that your implementation will be storing Taffy-specific data, like style and layout information. For the `Flexbox` algorithm, Taffy expects that your implementation of this trait will provide Flexbox-specific data, like its flex rules and specified dimensions.
```

File: src/tree.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
+//! The baseline requirements of any UI Tree so Taffy can efficiently calculate the layout
+
+use crate::{
+    layout::{Cache, Layout},
+    prelude::*,
+};
+
+/// Any item that implements the LayoutTree can be layed out using Taffy's algorithms.
+///
+/// Generally, Taffy expects your Node tree to be indexable by stable indicies. A "stable" index means that the Node's ID
```
Comment (alice-i-cecile) [inline src/tree.rs:10]:
```suggestion
/// Generally, Taffy expects your Node tree to be indexable by stable indices. A "stable" index means that the Node's ID
```

**Review by alice-i-cecile — APPROVED** [review]
I'm on board. This is well done, and in-line with my broader vision for how this library should evolve to support multiple layouts.

Merging now.

**Comment by nicoburns** [comment]
Just wanted to say thanks for reviewing this :)

**Pull request merged**